tests/unit/Storefront/Controller/ContextGatewayControllerTest.php

```diff
@@ -0,0 +1,159 @@
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\Session\FlashBagAwareSessionInterface;
+use Symfony\Component\HttpFoundation\Session\Session;
+
+/**
+ * @internal
+ */
+#[Package('framework')]
+#[CoversClass(ContextGatewayController::class)]
+class ContextGatewayControllerTest extends TestCase
+{
+    public function testGateway(): void
+    {
+        $context = Generator::generateSalesChannelContext(token: 'hatoken');
+        $cart = new Cart('hatoken');
+        $request = new Request(request: ['foo' => 'bar', 'bat' => 'baz']);
+        $expectedResponse = new ContextTokenResponse('newHatoken');
+
+        $cartService = $this->createMock(CartService::class);
+        $cartService
+            ->expects($this->once())
+            ->method('getCart')
+            ->with('hatoken', $context)
+            ->willReturn($cart);
+
+        $route = $this->createMock(AbstractContextGatewayRoute::class);
+        $route
+            ->expects($this->once())
+            ->method('load')
+            ->with($request, $cart, $context)
+            ->willReturn($expectedResponse);
+
+        $container = $this->createStubContainerWithFlashBag();
+
+        $controller = new ContextGatewayController($route, $cartService);
+        $controller->setContainer($container);
+
+        $newResponse = $controller->gateway($request, $context);
+
+        static::assertInstanceOf(ContextTokenResponse::class, $newResponse);
+        static::assertEquals('newHatoken', $newResponse->getToken());
+    }
+
+    public function testGatewayWithGenericException(): void
+    {
+        $context = Generator::generateSalesChannelContext(token: 'hatoken');
+        $cart = new Cart('hatoken');
+        $request = new Request(request: ['foo' => 'bar', 'bat' => 'baz']);
+
+        $cartService = $this->createMock(CartService::class);
+        $cartService
+            ->expects($this->once())
+            ->method('getCart')
+            ->with('hatoken', $context)
+            ->willReturn($cart);
+
+        $route = $this->createMock(AbstractContextGatewayRoute::class);
+        $route
+            ->expects($this->once())
+            ->method('load')
+            ->with($request, $cart, $context)
+            ->willThrowException(GatewayException::emptyAppResponse('FOO'));
+
+        $container = $this->createStubContainerWithFlashBag();
+
+        $controller = new ContextGatewayController($route, $cartService);
+        $controller->setContainer($container);
+
+        $newResponse = $controller->gateway($request, $context);
+
+        static::assertInstanceOf(JsonResponse::class, $newResponse);
+        static::assertEquals(Response::HTTP_BAD_REQUEST, $newResponse->getStatusCode());
+
+        /** @var FlashBagAwareSessionInterface $session */
```

Comment:
please do not use var annotations. Especially in tests you can simply assert the type. also see https://github.com/shopware/shopware/blob/trunk/coding-guidelines/core/writing-code-for-static-analysis.md